File: frontend/src/lib/analyzer.ts
import githubData from '@/data/github_data.json';

interface ContributorData {
  name: string;
  avatar_url: string;
  prs_created: number;
  total_files_changed: number;
  total_additions: number;
  total_deletions: number;
  avg_time_to_merge_hours: number;
  reviews_given: number;
  prs_reviewed: number;
}

interface PRData {
  merged_at: string;
  created_at: string;
  [key: string]: unknown;
}

interface EngineerScore {
  username: string;
  name: string;
  avatar_url: string;
  impact_score: number;
  quality_score: number;
  velocity_score: number;
  collaboration_score: number;
  leadership_score: number;
  stats: {
    prs_created: number;
    reviews_given: number;
    files_changed: number;
    avg_merge_time: number;
  };
}

const contributors = githubData.contributors as Record<string, ContributorData>;
const prs = (githubData.prs || []) as PRData[];

/** Calculate the number of months the data spans, based on PR merged dates */
function getDataWindowMonths(): number {
  if (prs.length === 0) return 3; // fallback
  const dates = prs
    .map(pr => new Date(pr.merged_at).getTime())
    .filter(d => !isNaN(d));
  if (dates.length === 0) return 3;
  const earliest = Math.min(...dates);
  const latest = Math.max(...dates);
  const months = (latest - earliest) / (1000 * 60 * 60 * 24 * 30);
  return Math.max(months, 1); // at least 1 month
}

/** Get the date range of the data as formatted strings */
export function getDataDateRange(): { from: string; to: string; months: number } {
  if (prs.length === 0) return { from: '', to: '', months: 0 };
  const dates = prs
    .map(pr => new Date(pr.merged_at).getTime())
    .filter(d => !isNaN(d));
  if (dates.length === 0) return { from: '', to: '', months: 0 };
  const earliest = new Date(Math.min(...dates));
  const latest = new Date(Math.max(...dates));
  const months = Math.round((latest.getTime() - earliest.getTime()) / (1000 * 60 * 60 * 24 * 30));
  const fmt = (d: Date) => d.toLocaleDateString('en-US', { month: 'short', year: 'numeric' });
  return { from: fmt(earliest), to: fmt(latest), months: Math.max(months, 1) };
}

const dataWindowMonths = getDataWindowMonths();

function calculateQualityScore(c: ContributorData): number {
  // Fast merge time = quality (inverse relationship)
  // Cap at 72 hours (3 days)
  const mergeTime = Math.min(c.avg_time_to_merge_hours, 72);
  const mergeScore = 100 * (1 - mergeTime / 72);

  // Reasonable PR size (sweet spot around 200-500 lines)
  const avgChanges = (c.total_additions + c.total_deletions) / Math.max(c.prs_created, 1);
  let sizeScore: number;
  if (avgChanges >= 200 && avgChanges <= 500) {
    sizeScore = 100;
  } else if (avgChanges < 200) {
    sizeScore = 50 + (avgChanges / 200) * 50;
  } else {
    sizeScore = Math.max(50, 100 - (avgChanges - 500) / 20);
  }

  // Active reviewer = understands quality
  const reviewActivity = Math.min(c.reviews_given / 20, 1) * 100;

  return 0.50 * mergeScore + 0.30 * sizeScore + 0.20 * reviewActivity;
}

function calculateVelocityScore(c: ContributorData): number {
  const prCount = c.prs_created;
  const filesChanged = c.total_files_changed;

  // Consistency: PRs per month (dynamically calculated from data window)
  const prsPerMonth = prCount / dataWindowMonths;
  const consistencyScore = Math.min(prsPerMonth / 10, 1) * 100;

  // Complexity: avg files per PR
  const avgFiles = filesChanged / Math.max(prCount, 1);
  const complexityScore = Math.min(avgFiles / 15, 1) * 100;

  return 0.40 * consistencyScore + 0.60 * complexityScore;
}

function calculateCollaborationScore(c: ContributorData): number {
  const reviewsGiven = c.reviews_given;
  const prsReviewed = c.prs_reviewed;

  // Review volume
  const reviewVolume = Math.min(reviewsGiven / 30, 1) * 100;

  // Review depth (reviews per unique PR)
  let reviewDepth = 0;
  if (prsReviewed > 0) {
    reviewDepth = Math.min(reviewsGiven / prsReviewed / 3, 1) * 100;
  }

  return 0.70 * reviewVolume + 0.30 * reviewDepth;
}

function calculateLeadershipScore(c: ContributorData): number {
  const filesChanged = c.total_files_changed;
  const prsCreated = c.prs_created;
  const reviewsGiven = c.reviews_given;

  // Code ownership (lots of files touched)
  const ownership = Math.min(filesChanged / 200, 1) * 100;

  // Dual role: both author AND reviewer
  const isAuthor = prsCreated > 0;
  const isReviewer = reviewsGiven > 0;
  let balance = 0;
  if (isAuthor && isReviewer) {
    balance = Math.min((prsCreated + reviewsGiven) / 40, 1) * 100;
  }

  return 0.60 * ownership + 0.40 * balance;
}

function calculateImpactScore(username: string): EngineerScore {
  const c = contributors[username];

  const quality = calculateQualityScore(c);
  const velocity = calculateVelocityScore(c);
  const collaboration = calculateCollaborationScore(c);
  const leadership = calculateLeadershipScore(c);

  const impact = 0.30 * quality + 0.30 * velocity + 0.20 * collaboration + 0.20 * leadership;

  return {
    username,
    name: c.name,
    avatar_url: c.avatar_url,
    impact_score: Math.round(impact * 10) / 10,
    quality_score: Math.round(quality * 10) / 10,
    velocity_score: Math.round(velocity * 10) / 10,
    collaboration_score: Math.round(collaboration * 10) / 10,
    leadership_score: Math.round(leadership * 10) / 10,
    stats: {
      prs_created: c.prs_created,
      reviews_given: c.reviews_given,
      files_changed: c.total_files_changed,
      avg_merge_time: c.avg_time_to_merge_hours,
    },
  };
}

export function getAllEngineers(): EngineerScore[] {
  const scores: EngineerScore[] = [];

  for (const username of Object.keys(contributors)) {
    const c = contributors[username];
    if (c.prs_created < 2) continue;
    scores.push(calculateImpactScore(username));
  }

  scores.sort((a, b) => b.impact_score - a.impact_score);
  return scores;
}

export function getTopEngineers(limit: number = 5): EngineerScore[] {
  return getAllEngineers().slice(0, limit);
}

export function getMethodology() {
  return {
    overview:
      'Impact measured across four dimensions with weighted scoring. Avoids vanity metrics like lines of code.',
    dimensions: [
      {
        name: 'Code Quality',
        weight: 0.3,
        description: 'Fast merge time, reasonable PR size, review activity',
        signals: [
          'Merge efficiency (lower time = higher quality)',
          'PR size optimization (200-500 lines sweet spot)',
          'Review engagement (understands quality standards)',
        ],
      },
      {
        name: 'Delivery Velocity',
        weight: 0.3,
        description: 'Consistent delivery of complex work',
        signals: [
          'PRs per month (consistency)',
          'Files changed per PR (complexity handling)',
        ],
      },
      {
        name: 'Collaboration',
        weight: 0.2,
        description: 'Helping teammates through code reviews',
        signals: [
          'Reviews given (volume)',
          'Comments per PR reviewed (depth)',
        ],
      },
      {
        name: 'Technical Leadership',
        weight: 0.2,
        description: 'Code ownership and balanced contributions',
        signals: [
          'Files touched (ownership breadth)',
          'Both authoring and reviewing (technical authority)',
        ],
      },
    ],
    philosophy:
      "This approach resists gaming. You can't just spam commits, make tiny PRs, or rubber-stamp reviews. Real impact requires quality code, consistent delivery, helpful reviews, and technical ownership.",
  };
}

File: backend/analyzer.py
import json
from typing import Dict, List

class ImpactAnalyzer:
    def __init__(self, data_file="data/github_data.json"):
        with open(data_file) as f:
            self.data = json.load(f)
        self.contributors = self.data["contributors"]
        self.prs = self.data["prs"]
    
    def _calculate_quality_score(self, contributor: Dict) -> float:
        """
        Code Quality = How good is their code?
        
        Signals:
        - Fast merge time (high quality gets approved quickly)
        - Reasonable PR size (not too small/large)
        - Active in reviews (understand quality standards)
        """
        
        # Fast merge time = quality (inverse relationship)
        # Cap at 72 hours (3 days) - anything longer is outlier
        merge_time = min(contributor["avg_time_to_merge_hours"], 72)
        merge_score = 100 * (1 - merge_time / 72)  # Lower time = higher score
        
        # Reasonable PR size (sweet spot around 200-500 lines)
        # Too small = trivial, too large = hard to review
        avg_changes = (contributor["total_additions"] + contributor["total_deletions"]) / max(contributor["prs_created"], 1)
        if 200 <= avg_changes <= 500:
            size_score = 100
        elif avg_changes < 200:
            size_score = 50 + (avg_changes / 200) * 50  # 0-200 maps to 50-100
        else:  # > 500
            size_score = max(50, 100 - (avg_changes - 500) / 20)  # Decays after 500
        
        # Active reviewer = understands quality
        review_activity = min(contributor["reviews_given"] / 20, 1) * 100  # Cap at 20 reviews
        
        # Weighted combination
        quality = (
            0.50 * merge_score +      # Merge time is strongest signal
            0.30 * size_score +        # PR size matters
            0.20 * review_activity     # Review activity shows engagement
        )
        
        return quality
    
    def _calculate_velocity_score(self, contributor: Dict) -> float:
        """
        Delivery Velocity = How much meaningful work do they ship?
        
        NOT just volume - look at consistency + complexity
        """
        
        # PR count (but normalized, so not just raw volume)
        pr_count = contributor["prs_created"]
        
        # Files touched (complexity proxy)
        files_changed = contributor["total_files_changed"]
        
        # Consistency: PRs per month (assuming 90 day window)
        prs_per_month = pr_count / 3
        consistency_score = min(prs_per_month / 10, 1) * 100  # 10+ PRs/month = 100
        
        # Complexity: avg files per PR
        avg_files = files_changed / max(pr_count, 1)
        complexity_score = min(avg_files / 15, 1) * 100  # 15+ files/PR = 100
        
        velocity = (
            0.40 * consistency_score +   # Consistent delivery
            0.60 * complexity_score       # Handle complex changes
        )
        
        return velocity
    
    def _calculate_collaboration_score(self, contributor: Dict) -> float:
        """
        Collaboration = How much do they help the team?
        
        Signals:
        - Reviews given (helping others)
        - Breadth of reviews (not just one person)
        """
        
        reviews_given = contributor["reviews_given"]
        prs_reviewed = contributor["prs_reviewed"]
        
        # Review volume
        review_volume = min(reviews_given / 30, 1) * 100  # 30+ reviews = 100
        
        # Review depth (reviews per unique PR)
        # High ratio = deep reviews (multiple comments per PR)
        # Low ratio = shallow reviews (one comment per PR)
        if prs_reviewed > 0:
            review_depth = min(reviews_given / prs_reviewed / 3, 1) * 100  # 3+ comments/PR = 100
        else:
            review_depth = 0
        
        collaboration = (
            0.70 * review_volume +   # Volume matters
            0.30 * review_depth      # But depth matters too
        )
        
        return collaboration
    
    def _calculate_leadership_score(self, contributor: Dict) -> float:
        """
        Technical Leadership = Are they a go-to expert?
        
        Signals:
        - Own significant code (high file count)
        - Both author AND reviewer (technical authority)
        - Handle large changes (architectural work)
        """
        
        files_changed = contributor["total_files_changed"]
        prs_created = contributor["prs_created"]
        reviews_given = contributor["reviews_given"]
        
        # Code ownership (lots of files touched)
        ownership = min(files_changed / 200, 1) * 100  # 200+ files = 100
        
        # Dual role: both author AND reviewer (shows breadth)
        is_author = prs_created > 0
        is_reviewer = reviews_given > 0
        if is_author and is_reviewer:
            balance = min((prs_created + reviews_given) / 40, 1) * 100  # 40+ total = 100
        else:
            balance = 0  # Must do both
        
        leadership = (
            0.60 * ownership +    # File ownership is key
            0.40 * balance        # Balanced contribution
        )
        
        return leadership
    
    def calculate_impact_score(self, username: str) -> Dict:
        """Calculate all dimension scores for a contributor"""
        contributor = self.contributors[username]
        
        quality = self._calculate_quality_score(contributor)
        velocity = self._calculate_velocity_score(contributor)
        collaboration = self._calculate_collaboration_score(contributor)
        leadership = self._calculate_leadership_score(contributor)
        
        # Overall impact (weighted)
        impact = (
            0.30 * quality +
            0.30 * velocity +
            0.20 * collaboration +
            0.20 * leadership
        )
        
        return {
            "username": username,
            "name": contributor["name"],
            "avatar_url": contributor["avatar_url"],
            "impact_score": round(impact, 1),
            "quality_score": round(quality, 1),
            "velocity_score": round(velocity, 1),
            "collaboration_score": round(collaboration, 1),
            "leadership_score": round(leadership, 1),
            # Raw stats for reference
            "stats": {
                "prs_created": contributor["prs_created"],
                "reviews_given": contributor["reviews_given"],
                "files_changed": contributor["total_files_changed"],
                "avg_merge_time": contributor["avg_time_to_merge_hours"]
            }
        }
    
    def get_all_engineers(self) -> List[Dict]:
        """Get all engineers ranked by impact"""
        scores = []
        for username in self.contributors:
            contributor = self.contributors[username]
            if contributor["prs_created"] < 2:
                continue
            score_data = self.calculate_impact_score(username)
            scores.append(score_data)
        scores.sort(key=lambda x: x["impact_score"], reverse=True)
        return scores

    def get_top_engineers(self, limit: int = 5) -> List[Dict]:
        """Get top N engineers ranked by impact"""
        return self.get_all_engineers()[:limit]
